Optimized partial and smooth Bézier curve computations in `bezier.py`

chopan050 opened the pull request:
## Overview: What does this pull request change?

I made many changes to the functions defined in bezier.py to calculate Bézier curves faster. Here is a list of those changes, starting from the most important ones:
- Replaced `partial_quadratic_bezier_points`, `split_quadratic_bezier`, `subdivide_quadratic_bezier` and `quadratic_bezier_remap` with generalized and **optimized** analogs `partial_bezier_points`, `split_bezier`, `subdivide_bezier` and `bezier_remap`
  - `bezier_remap` reused `(OpenGL)VMobject.insert_n_curves_to_point_list`, so I rewrote the latter methods by using `bezier_remap` instead
- Optimized `get_smooth_cubic_bezier_handle_points`
  - Redirected `get_smooth_handle_points` to `get_smooth_cubic_bezier_handle_points`
  - Added `get_smooth_cubic_bezier_handle_points_for_open_curve` and `get_smooth_cubic_bezier_handle_points_for_closed_curve`
  - Removed the auxiliar function `diag_to_matrix` and the `scipy` import
- Greatly optimized `is_close`, and slightly optimized `bezier` and `interpolate`
<!--changelog-end-->

## Motivation and Explanation: Why and how do your changes improve the library?

### partial_bezier_points
The original algorithm made a lot of calls to `bezier`, making multiple redundant intermediate calculations that could've been reused for the other Béziers. Instead, in the n-th degree case I calculate all of the necessary points in a single pass, improving from O(n³) to O(n²) time. Even more, the process can be encoded as a matrix multiplication, which is what I did for the simple cases (up to cubic curves), giving around 10x speedup.

```py
def partial_bezier_points(points, a, b):
    # Border cases...
    
    degree = points.shape[0] - 1
    if degree == 3:
        # calculate portion_matrix..
        return portion_matrix @ points
    # if degree in (2, 1, 0)...

    # Fallback case for nth degree Béziers
    # It is convenient that np.array copies points
    arr = np.array(points)
    N = arr.shape[0]
    
    if a != 0:
        for i in range(1, N):
            arr[: N - i] += a * (arr[1 : N - i + 1] - arr[: N - i])

    if b != 1:
        if a != 0:
            mu = (1 - b) / (1 - a)
        else:
            mu = 1 - b
        for i in range(1, N):
            arr[i:] += mu * (arr[i - 1 : -1] - arr[i:])

    return arr
```

### split_bezier and subdivide_bezier
Again, the processes can be encoded as matrix multiplications. In the case of `subdivide_bezier`, these matrices can even be memoized, as they're always the same.

```py
def split_bezier(points, t):
    points = np.asarray(points)
    N, dim = points.shape
    degree = N-1
    if degree == 3:
        # calculate split_matrix...
        return split_matrix @ points
    # if degree in (1, 0)...
    
    # Fallback case for nth degree
    arr = np.empty((2, N, dim))
    arr[1] = points
    arr[0, 0] = points[0]

    for i in range(1, N):
        arr[1, : N - i] += t * (arr[1, 1 : N - i + 1] - arr[1, : N - i])
        arr[0, i] = arr[1, 0]

    return arr


def subdivide_bezier(points, n_divisions):
    points = np.asarray(points)
    if n_divisions == 1:
        return points

    N, dim = points.shape
    degree = N - 1
    if degree == 3:
        # get subdivision_matrix...
        return subdivision_matrix @ points
    # if degree in (2, 1, 0)...

    # Fallback case for nth degree
    beziers = np.empty((n_divisions, N, dim))
    beziers[-1] = points
    for curve_num in range(n_divisions - 1, 0, -1):
        curr = beziers[curve_num]
        prev = beziers[curve_num - 1]
        prev[0] = curr[0]
        for i in range(1, N):
            a = (n_divisions - curve_num - 1) / (n_divisions - curve_num)
            curr[: N - i] += a * (curr[1 : N - i + 1] - curr[: N - i])
            prev[i] = curr[0]

    return beziers.reshape(n_divisions * N, dim)
```

### `bezier_remap`
I realized that I could essentially reuse the code for `VMobject.insert_n_curves_to_point_list` for this. **Notice the use of `subdivide_bezier` instead of repeated use of `partial_bezier_points`, which is much faster than the original algorithm:**

```py
def bezier_remap(bezier_tuples, new_number_of_curves):
    bezier_tuples = np.asarray(bezier_tuples)
    current_number_of_curves, nppc, dim = bezier_tuples.shape
    repeat_indices = (
        np.arange(new_number_of_curves, dtype="i") * current_number_of_curves
    ) // new_number_of_curves

    split_factors = np.zeros(current_number_of_curves, dtype="i")
    np.add.at(split_factors, repeat_indices, 1)

    new_tuples = np.empty((new_number_of_curves, nppc, dim))
    index = 0
    for curve, sf in zip(bezier_tuples, split_factors):
        new_tuples[index : index + sf] = subdivide_bezier(curve, sf).reshape(
            sf, nppc, dim
        )
        index += sf
    return new_tuples
```

Because of the above, I also straight up rewrote `(OpenGL)VMobject.insert_n_curves_to_point_list` to use `bezier_remap` instead:
```py
def insert_n_curves_to_point_list(self, n: int, points: np.ndarray) -> np.ndarray:
    if len(points) == 1:
        nppcc = self.n_points_per_cubic_curve
        return np.repeat(points, nppcc * n, 0)
        
    bezier_tuples = self.get_cubic_bezier_tuples_from_points(points)
    new_number_of_curves = bezier_tuples.shape[0] + n
    new_bezier_tuples = bezier_remap(bezier_tuples, new_number_of_curves)
    new_points = new_bezier_tuples.reshape(-1, 3)
```

### get_smooth_handle_points
Now we arrive to the first function in `bezier.py` I modified, and the main reason I made this PR.

I had a test scene with more than 100 `ParametricFunction`s in it, which were all being updated. More than half of the render time was spent on `ParametricFunction.generate_points`, and one of the 3 main culprits was `VMobject.make_smooth`. There were many issues in that function I plan to address in another PR, but the one I'm going to address in this specific PR is `get_smooth_handle_points`.

The main bottleneck for this function is the call to `scipy.linalg.solve_banded`. **Apparently, it spends more time validating the arrays passed as a parameter, than actually solving the system of equations.**

**That's why I decided to manually implement the algorithm for solving these equations. But also, all the matrices follow the same pattern and their coefficients are pretty much always the same, varying only in size, so the diagonals can be hardcoded in the algorithm rather than stored explicitly. Some of them can even be memoized, which is what I did!**

In the end, I got a speedup of around 3x. I can show a Snakeviz output of this, but that output includes a change to the `is_closed` function which really needed a rewrite, so lemme talk about it first.

### is_closed

Repeatedly using `np.allclose` to compare only two 3D points each time is actually **very expensive** and takes a painful part of the total time of `get_smooth_handle_points`. **So I rewrote it to a simple "compare 1st coords, then compare 2nd coords, then compare 3rd coords", which resulted in almost a 30x speedup!**

```py
def is_closed(points: Point3D_Array) -> bool:
    start, end = points[0], points[-1]
    atol = 1e-8
    if abs(end[0] - start[0]) > atol:
        return False
    if abs(end[1] - start[1]) > atol:
        return False
    if abs(end[2] - start[2]) > atol:
        return False
    return True
```

SnakeViz output for `VMobject.make_smooth` before/after the changes to `get_smooth_handle_points` and `is_closed`:

| Before | After |
|--------|--------|
| ![image](https://github.com/ManimCommunity/manim/assets/49853152/377fad60-b830-4e25-aedd-9cceb34c055d) | ![image](https://github.com/ManimCommunity/manim/assets/49853152/09a1377e-b8e4-4bb8-ac0d-143bc3c9531e) |

A closer look to `get_smooth_handle_points`:

|Before|After|
|-|-|
|![image](https://github.com/ManimCommunity/manim/assets/49853152/1aed22eb-27b1-4618-9b33-19bfaaff7c51) | ![image](https://github.com/ManimCommunity/manim/assets/49853152/9cb16e02-4070-447c-92a8-2d8105b14201)|

The red block at the right is the new `is_closed`. It now takes much less time to run! Combined with the `get_smooth_handle_points` optimization, this latter function now has an overall speedup of around 4x.

### interpolate
Rewriting this function to use 1 product instead of 2, makes a speedup of around 30%:

```py
def interpolate(start, end, alpha):
    # return (1 - alpha) * start + alpha * end # original
    return start + alpha * (end - start) # new
```

### bezier
First of all, I added the cases for Bézier curves of grade 0 and 1. This was originally intended for `partial_bezier_curves`, which repeatedly called `bezier` and was not prepared for these specific cases, falling back to the expensive general nth grade case:

```py
def bezier(
    points: Sequence[Point3D] | Point3D_Array,
) -> Callable[[float | ColVector], Point3D | Point3D_Array]:
    P = np.asarray(points)
    num_points, dim = P.shape
    n = num_points - 1

    if n == 0:
        def zero_bezier(t):
            return np.ones_like(t) * P[0]
        return zero_bezier

    if n == 1:
        def linear_bezier(t):
            return P[0] + t * (P[1] - P[0])
        return linear_bezier

    # ...
```
Then, I made a very slight optimization for the cubic Bëzier case:
```py
    if n == 3:
        def cubic_bezier(t):
            t2 = t * t
            t3 = t2 * t
            mt = 1 - t
            mt2 = mt * mt
            mt3 = mt2 * mt
            return mt3 * P[0] + 3 * t * mt2 * P[1] + 3 * t2 * mt * P[2] + t3 * P[3]
        return cubic_bezier
```
It reported a small speedup of around 10% - 15%. Not too bad! I did something similar with the quadratic Bëzier:
```py
    if n == 2:
        def quadratic_bezier(t):
            t2 = t * t
            mt = 1 - t
            mt2 = mt * mt
            return mt2 * P[0] + 2 * t * mt * P[1] + t2 * P[2]
        return quadratic_bezier
```
Finally, the general nth grade Bézier case. I avoided the extensive use of the `choose` function which can be expensive, and instead used the recursive definition of Béziers I used for functions like `partial_bezier_points` and `subdivide_bezier`.

Also, `bezier` now returns named functions instead of lambdas, which is good for debugging and profiling: now you can know which function is taking what part of the time when profiling your code.

## Links to added or changed documentation pages
`manim.utils.bezier`
https://manimce--3281.org.readthedocs.build/en/3281/reference/manim.utils.bezier.html

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

chopan050 commented on manim/utils/bezier.py:
```diff
@@ -285,140 +521,339 @@
+        for i in range(len_memo, N - 1):
+            cp[i] = 1 / (4 - cp[i - 1])
+            up[i] = -cp[i] * up[i - 1]
+        CP_CLOSED_MEMO = cp
```
It's not used in the rest of the function, but its value is set for use in future calls of the function. Is there any way to disable this particular warning?

chopan050 replied:
Nvm, it's a notice, not a warning.

chopan050 commented on manim/utils/bezier.py:
```diff
@@ -90,35 +132,229 @@
+        arr[0] += a * (arr[1] - arr[0])  # arr = [Q0 L1 P2]
+
+    # Current state: arr = [Q0 L1 P2] = [P0' P1' P2']
+    if b != 0:
```
Fixed, I actually wanted to put `if b != 1:`. Oops!

chopan050 commented on manim/utils/bezier.py:
```diff
@@ -285,140 +521,339 @@
+            cp[i] = 1 / (4 - cp[i - 1])
+            up[i] = -cp[i] * up[i - 1]
+        CP_CLOSED_MEMO = cp
+        UP_CLOSED_MEMO = up
```
Same as before.

chopan050 commented on manim/utils/bezier.py:
```diff
@@ -285,140 +521,339 @@
+        cp[:len_memo] = CP_OPEN_MEMO
+        for i in range(len_memo, N - 1):
+            cp[i] = 1 / (4 - cp[i - 1])
+        CP_OPEN_MEMO = cp
```
Same as before.

chopan050 commented on manim/utils/bezier.py:
```diff
@@ -90,35 +132,229 @@
     end_prop = (b - a) / (1.0 - a)
     h1 = (1 - end_prop) * h0 + end_prop * h1_prime
     return [h0, h1, h2]
+    """
```
Fixed!

chopan050 commented on manim/utils/bezier.py:
```diff
@@ -90,35 +132,229 @@
     end_prop = (b - a) / (1.0 - a)
-    return np.array([bezier(a_to_1[: i + 1])(end_prop) for i in range(len(points))])
+    return np.array([bezier(a_to_1[: i + 1])(end_prop) for i in range(n+1)])
+    """
```
Fixed!

chopan050 commented:
Hello everyone! It's been a long time.

I made some updates and extra optimizations I discovered a long time ago while working on #3336 :)

chopan050 commented on manim/utils/bezier.py:
```diff
@@ -194,48 +196,534 @@
+    points = np.asarray(points)
+    N, dim = points.shape
+    degree = N - 1
+    tuple_shape = (2, N, dim)
```
I could swear I had removed that one... well, gonna fix that now.

MrDiver commented on manim/utils/bezier.py:
```diff
@@ -273,7 +760,7 @@ def interpolate(start: Point3D, end: Point3D, alpha: float) -> Point3D:
     start: int | float | Point3D, end: int | float | Point3D, alpha: float | Point3D
 ) -> float | Point3D:
-    return (1 - alpha) * start + alpha * end
+    return start + alpha * (end - start)
```
LOVE IT ❤️ GREAT JOB !!!!!!

chopan050 replied:
thx :heart:

MrDiver commented on manim/utils/bezier.py:
```diff
@@ -464,146 +951,408 @@ def match_interpolate(
-        else:
-            handle_pairs[:, i] = solve_func(b[:, i])
-    return handle_pairs[0::2], handle_pairs[1::2]
+# Figuring out which Bezier curves most smoothly connect a sequence of points
```
Figure out where comment was or just remove it if unnecessary

chopan050 replied:
Discussed in Discord, but I clarify for readers: in the end we dismissed this.
The confusion appeared because the diff was ruined.
But the comment is where it should belong.

MrDiver commented on manim/utils/bezier.py:
```diff
@@ -629,67 +1410,106 @@ def get_quadratic_approximation_of_cubic(
+    # this is a simple quadratic equation.
+    # However, if it's a 3D curve, then a, b and c are 3D vectors
+    # and this would require solving 3 quadratic equations.
+    # TODO: this simplifies by considering the 2D case. It might fail with 3D curves!
```
We need the 3D part, remove Todo

chopan050 replied:
Discussed in Discord, but I clarify for the readers: the original algorithm already assumed a 2D case and used the `cross2D` functions. I was not the one to make that simplification, I just documented better what was already there in the case of `get_quadratic_approximation_of_cubic`.
Therefore, this comment was dismissed. :+1:

MrDiver reviewed (changes requested):
Look over Docstrings
- [ ] Check that it is a human-readable format and indented formulas are either equations or monospaced format blocks like code blocks
- [x] Have renamed links with the `[]()` notation
- [x] Check that all the Todos are still necessary, and if so make a new issue for them so we can track them in the future, maybe also a combined issue for those.

behackl commented on manim/utils/bezier.py:
```diff
@@ -216,46 +211,538 @@ def quadratic_bezier_remap(
+
+    To understand the mathematics behind splitting curves, see split_bezier for an explanation.
+
+    To find the portion of a curve with t between a and b:
+    1. Split the curve at t = a and extract its 2nd subcurve.
+    2. We cannot evaluate the new subcurve at t = b because its range of values for t is different.
+       To find the correct value, we need to transform the interval [a, 1] into [0, 1]
+       by first subtracting a to get [0, 1-a] and then dividing by 1-a. Thus, our new
+       value must be t = (b - a) / (1 - a). Define u = (b-a) / (1-a).
+    3. Split the subcurve at t = u and extract its 1st subcurve.
```
This should fix the numbered list, did not test locally though.

```suggestion
    To find the portion of a curve with t between a and b:

    #. Split the curve at t = a and extract its 2nd subcurve.
    #. We cannot evaluate the new subcurve at t = b because its range of values for t is different.
       To find the correct value, we need to transform the interval [a, 1] into [0, 1]
       by first subtracting a to get [0, 1-a] and then dividing by 1-a. Thus, our new
       value must be t = (b - a) / (1 - a). Define u = (b-a) / (1-a).
    #. Split the subcurve at t = u and extract its 1st subcurve.
```

chopan050 replied:
Fixed!
In the end I just needed to add an extra blank line before starting the list. When Sphinx encounters groups of consecutive lines (non-separated by blank lines), it tends to put them in the same paragraph.

chopan050 commented on manim/utils/bezier.py:
```diff
@@ -216,46 +211,538 @@ def quadratic_bezier_remap(
+    return arr.reshape(2 * N, dim)
+
+
+CUBIC_SUBDIVISION_MATRICES = {
```
I essentially reused the `partial_matrix` from `partial_bezier_points` and replaced `a`, `b` with `(i-1)/n`, `i/n`. I added the respective docs.

MrDiver commented on manim/utils/bezier.py:
```diff
@@ -464,163 +1297,623 @@ def match_interpolate(
+            \end{pmatrix}
+
+        which will be expressed as :math:`TH_1 = D`.
+        :math:`T` is a tridiagonal matrix, so the system can be solved in :math`O(N)`
```
```suggestion
        :math:`T` is a tridiagonal matrix, so the system can be solved in :math:`O(N)`
```

chopan050 commented:
Update:

- Fixed `subdivide_bezier()` which had some typos causing wrong results for Béziers of degree 3 or higher
- Added tests for `partial_bezier_points()`, `split_bezier()` and `subdivide_bezier()`
